post.md

```diff
@@ -0,0 +1,166 @@
+
+This article and accompanying example application utilize Next.js version 14.x configured with TailwindCSS and TypeScript.
+
+## The Importance of Server-Side Validation
+
+While client-side validation provides immediate feedback and improves user experience, server-side validation is a must for any application. It acts as a second line of defense against invalid or malicious data. Even if you augment your application with client-side validation, server-side validation is necessary to ensure data integrity and security.
+
+## Overview of Zod Schema for Validation
+
+Zod allows us to define a validation schema for our form data. This schema is a declarative way of specifying the validation rules for each field. For example, to mark a field as required, we can use the `min(1)` method, which specifies that the field must have a minimum length of 1.
+
+Here's an example of a Zod schema definition:
+
+```ts
+import { z } from "zod";
+
+const schema = z.object({
+  name: z.string().min(1, "Name cannot be blank"),
+  email: z.string().email("Invalid email address"),
+});
+```
+
+## Handling Server-Side Validation Errors
+
+When we submit our form, the server-side action validates the form data against the Zod schema. If the validation fails, the server returns an array of errors (`ZodIssue[]`). Each error object has the following structure:
+
+```json
+{
+  "code": "too_small",
+  "minimum": 1,
+  "type": "string",
+  "inclusive": true,
+  "exact": false,
+  "message": "Name cannot be blank",
+  "path": ["name"]
+}
+```
+
+Now, let's look at how we can implement this in our Next.js application. We'll start with the server action, then move on to the contact page, and finally the form component.
+
+## Server Action
+
+In our server action, we validate the form data using the Zod schema. If the validation fails, we return the array of errors. If the validation succeeds, we redirect the user to the home page.
+
+```ts
+// app/contact/server-action.ts
+"use server";
+
+import { redirect } from "next/navigation";
+import { z } from "zod";
+
+const schema = z.object({
+  name: z.string().min(1, "Name cannot be blank"),
+  email: z.string().email().min(1, "Email cannot be blank"),
+});
+
+export default async function contactAction(_prevState: any, params: FormData) {
+  const validation = schema.safeParse({
+    name: params.get("name"),
+    email: params.get("email"),
+  });
+
+  if (validation.success) {
+    // save the data, send an email, etc.
+    redirect("/");
+  } else {
+    return {
+      errors: validation.error.issues,
+    };
+  }
+}
+```
+
+## Contact Page
+
+In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a server component while the `Form` component is a client component. We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
```

Comment:
I suggest adding doc links to the phrases _server component_ and _client component_.